Standalone ng generate schematic prerequisites and step one

crisbeto opened the pull request:
These changes add some of the prerequisites for the first step of the standalone `ng generate` schematic, as well as the first step itself. There's more info in the individual commits, but here's a high-level overview.

1. Allows aliases to be specified in the `ImportManager`.
2. Some minor improvements to our schematic utilities.
3. Boilerplate for `ng generate` schematics in `core`.
4. First mode of the `ng generate @angular/core:standalone` schematic. There will be two more that are going to be added in follow-up PRs.

crisbeto committed:
refactor(migrations): allow for aliases to be specified in the import manager

Adds the ability to set an alias when adding an import through the `ImportManager`. This will be required in some upcoming schematics.

crisbeto committed:
refactor(migrations): minor improvements to existing schematic utilties

Makes the following minor improvements to our current schematic utilities:
* `closestNode` now takes a TS predicate function instead of a `SyntaxKind`. This allows it automatically infer the type of the match.
* `getImportSpecifier` now accepts a regex for the module name. This will be useful for some upcoming migrations.
* Splits the logic for creating the migration program options into a separate function so that it's easier to reuse.
* `createMigrationProgram` now returns the program directly, instead of a literal with some other information.
* `FakeReadFileFn` and `createMigrationCompilerHost` aren't exported anymore since they aren't used anywhere.

crisbeto committed:
build: set up ng generate schematics for core

Adds the necessary boilerplate to allow for `ng generate` schematics to be included with `@angular/core`.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/README.md:
```diff
@@ -0,0 +1,2 @@
+# Standalone migration
+TODO
```
This will be added in one of the later commits.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/util.ts:
```diff
@@ -0,0 +1,251 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
```
There are a few utilties in this file that aren't being used yet. They'll come into play in the next PRs.

devversion commented on packages/core/schematics/utils/import_manager.ts:
```diff
@@ -103,11 +104,14 @@ export class ImportManager {
-                  e.propertyName ? e.propertyName.text === symbolName : e.name.text === symbolName);
+          const existingElement = namedBindings.elements.find(e => {
+            if (alias) {
+              return e.propertyName && e.name.text === alias && e.propertyName.text === symbolName;
```
Just to discuss this a little, so when an alias is explicitly specified, we will only use an existing import specifier if it's aliased to the _exact same_ name. 

It would be useful knowing (since I'm going commit per commit) why we _need_ to be sure that the alias is exactly the one provided, instead of reusing the existing arbitrary alias?

crisbeto replied:
In one of the later steps of the migration we need to copy code from one file into another. It's easier and safer to preserve the alias than to try and swap it out with the unique one everywhere where it's being referenced. I'm aware that this can cause a conflict if there's an import with the exact same alias in the new file, but at least for the migrations we have so far, that's extremely unlikely.

devversion replied:
I see. Should we leave a TODO to revisit this in the future when we e.g. would use the import manager in a more shared approach? I would want to make sure it's working correctly in terms of not generating invalid imports.

crisbeto replied:
Sure, I've left a TODO.

devversion commented on packages/core/schematics/utils/import_manager.ts:
```diff
@@ -301,6 +300,36 @@ export class ImportManager {
+
+    if (needsGeneratedUniqueName) {
+      propertyName = symbolIdentifier;
+      name = generatedUniqueIdentifier;
```
Especially here, if the alias would conflict in the source file it would be adjusted to be unique. e.g. `alias = bla` could become `{X from bla_1}`

devversion replied:
answered above, the alias support is not fully correct anyway.

devversion commented on packages/core/schematics/test/standalone_migration_spec.ts:
```diff
@@ -0,0 +1,1175 @@
+            `@NgModule({imports: [CommonModule, MyComp, MyOtherComp], exports: [MyComp]})`));
+  });
+
+  it('should add imports to pipes that are used in the template', async () => {
```
do we want a test case for the case where multiple pipes with the same name match (the case where get potential pipes returns >1)

crisbeto replied:
As I mentioned above, I think that'll be fairly rare.

devversion commented on packages/core/schematics/test/standalone_migration_spec.ts:
```diff
@@ -0,0 +1,1175 @@
+      @Component({
+        template: '<hello></hello>',
+        standalone: true,
+        imports: [ButtonModule, MatCardModule]
```
One question: Should `Hello` be also an import here?

crisbeto replied:
Technically it should, but I intentionally didn't include copying of test module `declarations`, because I thought that it has a high chance of being incorrect and breaking a test in a vague way. This way at least the error message should be pretty obvious. Since we don't have access to the type checker in test files, the migration is more of a "best attempt" rather than trying to be entirely correct.

devversion replied:
Sounds reasonable, good that we have chatted about this though.

devversion reviewed (approved):
Nice work! just a few questions/minor comments

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/to-standalone.ts:
```diff
@@ -0,0 +1,498 @@
+  } else if (ts.isObjectLiteralExpression(node.expression.arguments[0])) {
+    literalProperties = [...node.expression.arguments[0].properties, property];
+  } else {
+    // Unsupported case (e.g. `@Component(SOME_CONST)`). Return the original node.
```
My thinking was that these cases are pretty rare so it's not worth it to insert a TODO.

devversion replied:
Yeah, definitely rare. Just seems trivial to add a comment, and to avoid future issues where people would report some classes not being migrated. Your call

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/to-standalone.ts:
```diff
@@ -0,0 +1,498 @@
+  if (usedPipes !== null) {
+    const potentialPipes = typeChecker.getPotentialPipes(ref.node);
+
+    for (const pipe of potentialPipes) {
```
In the `if` condition below we only keep the pipes that are used. You're right that there might be multiple pipes with the same name, but those cases should be rare and I'm not totally sure how we would pick one. Even if we import two pipes with the same name, the user should get a pretty clear error to fix it themselves.

devversion replied:
Sounds reasonable. Yeah, having all of them referenced is likely more clear than not having any referenced.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/to-standalone.ts:
```diff
@@ -0,0 +1,498 @@
+
+      // Note that we only migrate classes that are in the same file as the testing module,
+      // because external fixture components are somewhat rare and handling them is going
+      // to involve a lot of assumptions that are likely to be incorrect.
```
I'll see how common it is during testing on a real project and re-evaluate whether it needs a TODO.

crisbeto committed:
feat(core): add ng generate schematic to convert declarations to standalone

Implements a new `ng generate @angular/core:standalone` schematic that allows the user to convert all the declarations in a set of NgModules to standalone.

crisbeto reviewed (commented):
Feedback has been addressed.

jessicajaniuk commented:
This PR was merged into the repository by commit a154db8a81cbdfed8c3d0db1e2a5bf43aa3e0bbf.
